### ISSUE TITLE
Replace state decorators

### ISSUE THREAD

**Issue by vicgeralds** [Lundalogik/lime-web-components#537]
## New feature motivation

TypeScript decorators were an experimental feature based on a stage 2 proposal for JavaScript that has now been superseded by a [stage 3 proposal](https://github.com/tc39/proposal-decorators) that's quite different. The new version of decorators will not allow us to decorate properties the way we have implemented state decorators today (overriding lifecycle methods). One major design goal is to prevent exactly what we're doing: "Decorators should affect the thing they're decorating, and avoid confusing/non-local effects." 

Maybe TypeScript will still continue to support decorators as they have been implemented, but moving forward I think we should try to avoid relying on them. More on this: https://www.reddit.com/r/typescript/comments/oa3x1x/do_decorators_have_a_future/

## New feature description

Instead of providing state decorators, we can provide functions for "decorating" properties by calling them in initializers.

```ts
class MyComponent implements LimeWebComponent {
  ...
  @State()
  private myState = selectState(this);
  ...
}
```

Another option is to provide React-like hooks. Some inspiration: https://github.com/saasquatch/stencil-hooks

Other ideas?

**Comment by adrianschmidt** [comment]
> Instead of providing state decorators, we can provide functions for "decorating" properties by calling them in initializers.
> 
> ```ts
> class MyComponent implements LimeWebComponent {
>   ...
>   @State()
>   private myState = selectState(this);
>   ...
> }
> ```

YES PLEASE! This is very straightforward and easy to understand! Love it!

> Another option is to provide React-like hooks. Some inspiration: [saasquatch/stencil-hooks](https://github.com/saasquatch/stencil-hooks)

I didn't feel like I had the time to look too far into it, but the fact that I didn't understand what's going on at first glance, and with your suggestion above I _did_, I don't feel like I need to, or want to, look any further into hooks. Simple is good! ❤️

**Comment by vicgeralds** [comment]
> I don't feel like I need to, or want to, look any further into hooks.

If we can keep simple things this simple, I think that's great. As long as state is an object (or array) we can just update it and force render instead of setting the property with decorator magic.

When we need more advanced subscriptions, or state that should be updated based on props or other state, I hope we can have something that's nicer than this though:

```ts
class MyComponent implements LimeWebComponent {
  @Prop()
  public myValue: Expression;
  
  @State()
  private otherState = 'default';

  @State()
  private myState = selectState(this, this.myValue, this.otherState);

  @Watch('context')
  contextChanged() {
    this.myState.unsubscribe();
    this.myState = selectState(this, this.myValue, this.otherState);
  }
  
  @Watch('otherState')
  otherStateChanged() {
    this.myState.unsubscribe();
    this.myState = selectState(this, this.myValue, this.otherState);
  }
}
```

Maybe we could get rid of the watchers and just do:

```ts
componentWillRender() {
  this.myState.update(this, this.myValue, this.otherState);
}
```

With hooks, you would instead do something like this:

```ts
class MyComponent implements LimeWebComponent {
  @Prop()
  public myValue: Expression;
  
  @State()
  private otherState = 'default';

  render() {
    const myState = useLimeService(this, this.myValue, this.otherState);
    ...
  }
}
```

But outside React, you would also need to enable hooks on the component somehow, for example in the constructor like they do in stencil-hooks

```ts
constructor() {
  withHooks(this);
}
```

but I would prefer not exposing the concept of hooks like that, even if would expose a React-like API for those familiar with it.
